File: QytetetJava/src/modeloqytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package modeloqytetet;

/**
 *
 * @author felix
 */
public class Calle extends Casilla{
    
    protected int coste;
    protected int numHoteles;
    protected int numCasas;
    protected TituloPropiedad titulo;

    public Calle(int numeroCasilla, int coste, TipoCasilla tipo, TituloPropiedad titulo) {
        super(numeroCasilla, tipo);
        this.coste = coste;
        this.titulo = titulo;
        titulo.setCasilla(this);
        this.numHoteles = 0;
        this.numCasas = 0;
        if (tipo == TipoCasilla.IMPUESTO)
            this.coste = 300;
    }
    
    public void setTitulo(TituloPropiedad titulo) {
        this.titulo = titulo;
    }

    public void setNumHoteles(int numHoteles) {
        this.numHoteles = numHoteles;
    }

    public void setNumCasas(int numCasas) {
        this.numCasas = numCasas;
    }
    
    public TituloPropiedad getTitulo() {
        return titulo;
    }
    
    TituloPropiedad asignarPropietario(Jugador jugador) {
        titulo.setPropietario(jugador);    
        return titulo;
    }
    
    int calcularValorHipoteca() {
        int valorHipotecaBase = titulo.getHipotecaBase();
        int total = (int) (valorHipotecaBase + numCasas * 0.5 * valorHipotecaBase + numHoteles * valorHipotecaBase);
        
        return total;
    }
    
    int cancelarHipoteca() {
        int pago = 0;
        
        if (this.estaHipotecada()) {
           titulo.setHipotecada(false);
           pago = (int) ( this.calcularValorHipoteca() * 1.1);
       }
        
        return pago;
    }
    
    int cobrarAlquiler() {
        int costeAlquilerBase = titulo.getAlquilerBase();
        int costeAlquiler = costeAlquilerBase + (int)(numCasas * 0.5 + numHoteles * 2);
        
        titulo.cobrarAlquiler(costeAlquiler);
        
        return costeAlquiler;
    }
    
    int edificarCasa() {
        int costeEdificarCasa;
        this.setNumCasas(numCasas + 1);
        
        costeEdificarCasa = titulo.getPrecioEdificar();
        
        return costeEdificarCasa;
    }
    
    int edificarHotel() {
        int costeEdificarHotel;
        this.setNumHoteles(numHoteles + 1);
        
        costeEdificarHotel = titulo.getPrecioEdificar();
        
        return costeEdificarHotel;
    }
    
    boolean estaHipotecada() {
        return titulo.getHipotecada();
    }
    
    int getCoste() {
        return coste;
    }
    
    int getCosteHipoteca() {
        return titulo.getHipotecaBase();
    }
    
    int getNumCasas() {
        return numCasas;
    }
    
    int getNumHoteles() {
        return numHoteles;
    }
    
    int getPrecioEdificar() {
        return titulo.getPrecioEdificar();
    }
    
    int hipotecar() {
        titulo.setHipotecada(true);
        int hipotecaBase = this.calcularValorHipoteca();
        int cantidadRecibida = (int) (hipotecaBase + this.getNumCasas() * 0.5 * hipotecaBase + this.getNumHoteles() * hipotecaBase);
        
        return cantidadRecibida;
    }
    
    int precioTotalComprar() {
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    boolean propietarioEncarcelado() {
        return titulo.propietarioEncarcelado();
    }
    
    boolean sePuedeEdificarCasa(int factorEspeculador){
        int cuantas = 4 * factorEspeculador; 
        return numCasas < cuantas;
    }
    
    boolean sePuedeEdificarHotel(int factorEspeculador){
         int cuantasCasas = 4 * factorEspeculador; 
         int cuantosHoteles = 4 * factorEspeculador; 
         return numCasas == cuantasCasas && numHoteles < cuantosHoteles;
    }
     
    boolean tengoPropietario() {
        return titulo.tengoPropietario();
    }
    
    int venderTitulo() {
        float precioVenta;
        int precioCompra;
        
        titulo.setPropietario(null);
        this.setNumHoteles(0);
        this.setNumCasas(0);
        
        precioCompra = coste + (numCasas+numHoteles) * titulo.getPrecioEdificar();
        precioVenta = precioCompra + titulo.getFactorRevalorizacion() * precioCompra;
        
        return (int)precioVenta;
    }
    
    void asignarTituloPropiedad() {
        if(this.tipo == TipoCasilla.CALLE)
            this.titulo.setCasilla(this);
    }  
    
    @Override
    public String toString() {
        return "\nCasilla{" + "numeroCasilla=" + numeroCasilla + ", coste=" + coste + ", numHoteles=" + numHoteles + ", numCasas=" + numCasas + ", tipo=" + tipo + ", titulo=" + titulo + '}';
    }
    
}

File: QytetetRuby/lib/calle.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "tablero"
require_relative "jugador"
require_relative "titulo_propiedad"
require_relative "tipo_casilla"
require_relative "casilla"


module ModeloQytetet

  class Calle 
    attr_accessor :numeroCasilla , :tipo , :coste ,:numCasas, :numHoteles, :titulo
    
    def initialize(numeroCasilla, coste, tipo, titulo)
      @numeroCasilla = numeroCasilla
      @coste = coste
      @numHoteles = 0
      @numCasas = 0
      @titulo = titulo 
      @tipo = tipo
      if (tipo == TipoCasilla::IMPUESTO) then
        @coste = 300
      end
    end
    
     def set_titulo(titulo)
      @titulo = titulo
    end
    
    def set_num_hoteles(nuevoNumero)
      @numHoteles = nuevoNumero
    end
    
    def set_num_casas(nuevoNumero)
      @numCasas = nuevoNumero
    end
    
    def get_titulo
      @titulo
    end
    
    def asignar_propietario(jugador)
      @titulo.propietario = jugador
      
      @titulo
    end
    
    def calcular_valor_hipoteca
      valor_hipoteca_base = @titulo.hipotecaBase
      total = valor_hipoteca_base + @numCasas * 0.5 * valor_hipoteca_base + @numHoteles * valor_hipoteca_base
      
      total
    end
    
    def cancelar_hipoteca
      pago = 0
      
      if esta_hipotecada then
        @titulo.hipotecada = false
        pago = calcular_valor_hipoteca * 1.1
      end
      
      pago
    end
    
    def cobrar_alquiler
      coste_alquiler_base = @titulo.alquilerBase
      coste_alquiler = coste_alquiler_base + @numCasas * 0.5 + @numHoteles * 2
    
      @titulo.cobrar_alquiler(coste_alquiler)
      
      coste_alquiler
    end
    
    def edificar_casa
      set_num_casas(@numCasas + 1)
      
      coste_edificar_casa = @titulo.precioEdificar

      coste_edificar_casa
    end
    
    def edificar_hotel
      set_num_hoteles(@numHoteles + 1)
      
      coste_edificar_hotel = @titulo.precioEdificar
      
      coste_edificar_hotel
    end
    
    def esta_hipotecada
      @titulo.get_hipotecada
    end
    
    def get_coste
      @coste
    end
    
    def get_coste_hipoteca
      @titulo.hipotecaBase
    end
    
    def get_num_casas
      @numCasas
    end
    
    def get_num_hoteles
      @numHoteles
    end
    
    def get_precio_edificar
      @titulo.get_precio_edificar
    end
    
    def hipotecar
      @titulo.set_hipotecada(true)
      hipoteca_base = calcular_valor_hipoteca
      cantidad_recibida = hipoteca_base + @numCasas * 0.5 * hipoteca_base + @numHoteles * hipoteca_base
    
      cantidad_recibida
    end
    
    def precio_total_comprar
      
    end
    
    def propietario_encarcelado
      @titulo.propietario_encarcelado
    end
    
    def se_puede_edificar_casa(factorEspeculador)
      cuantas = 4 * factorEspeculador
      if @numCasas < cuantas then 
        return true
      else
        return false
      end 
    end
    
    def se_puede_edificar_hotel(factorEspeculador)
      cuantas_casas = 4 * factorEspeculador
      cuantos_hoteles = 4 * factorEspeculador
      return @numCasas == cuantas_casas && @numHoteles < cuantos_hoteles
    end
    
    def tengo_propietario
      @titulo.tengo_propietario
    end
    
    def vender_titulo
      @titulo.set_propietario(nil)
      set_num_hoteles(0)
      set_num_casas(0)
      
      precio_compra = @coste + (@numCasas + @numHoteles) * @titulo.precioEdificar
      precio_venta = precio_compra + @titulo.factorRevalorizacion * precio_compra
    
      precio_venta
    end
    
    def asignar_titulo_propiedad
      if @tipo == TipoCasilla::CALLE then
        @titulo.casilla = self
      end
    end
    
    def get_tipo
      @tipo
    end
    
    def get_numero_casilla
      @numeroCasilla
    end
    
    def soy_edificable
      @tipo == TipoCasilla::CALLE
    end
    
    def to_s
      "Calle\n { numeroCasilla= #{@numeroCasilla}, coste= #{@coste}, numHoteles= #{@numHoteles}, numCasas= #{@numCasas}, tipo= #{@tipo}, titulo= #{@titulo} }\n"
    end
    
    private :set_titulo
  end
  
end
